src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultActionDescriptorCollectionProvider.cs

```diff
@@ -0,0 +1,160 @@
+                .OrderBy(p => p.Order)
+                .ToArray();
+
+            _actionDescriptorChangeProviders = actionDescriptorChangeProviders.ToArray();
+
+            ChangeToken.OnChange(
+                GetCompositeChangeToken,
+                UpdateCollection);
+
+            _lock = new object();
+        }
+
+        /// <summary>
+        /// Returns a cached collection of <see cref="ActionDescriptor" />.
+        /// </summary>
+        public override ActionDescriptorCollection ActionDescriptors
+        {
+            get
+            {
+                Initialize();
+                Debug.Assert(_collection != null);
+                Debug.Assert(_changeToken != null);
+
+                return _collection;
+            }
+        }
+
+        /// <summary>
+        /// Gets an <see cref="IChangeToken"/> that will be signaled after the <see cref="ActionDescriptors"/>
+        /// collection has changed.
+        /// </summary>
+        /// <returns>The <see cref="IChangeToken"/>.</returns>
+        public override IChangeToken GetChangeToken()
+        {
+            Initialize();
+            Debug.Assert(_collection != null);
+            Debug.Assert(_changeToken != null);
+
+            return _changeToken;
+        }
+
+        private IChangeToken GetCompositeChangeToken()
+        {
+            if (_actionDescriptorChangeProviders.Length == 1)
+            {
+                return _actionDescriptorChangeProviders[0].GetChangeToken();
+            }
+
+            var changeTokens = new IChangeToken[_actionDescriptorChangeProviders.Length];
+            for (var i = 0; i < _actionDescriptorChangeProviders.Length; i++)
+            {
+                changeTokens[i] = _actionDescriptorChangeProviders[i].GetChangeToken();
+            }
+
+            return new CompositeChangeToken(changeTokens);
+        }
+        
+        private void Initialize()
+        {
+            // Using double-checked locking on initialization because we fire change token callbacks
+            // when the collection changes. We don't want to do that repeatedly for redundant changes.
+            //
+            // The main call path of this code on the first call is async initialization from Endpoint Routing
+            // which is done in a non-blocking way so in practice no caller will ever block here.
+            if (_collection == null)
+            {
+                lock (_lock)
+                {
+                    if (_collection == null)
+                    {
+                        UpdateCollection();
+                    }
+                }
+            }
+        }
+
+        private void UpdateCollection()
+        {
+            // Using the lock to initialize writes means that we serialize changes. This eliminates
+            // the potential for changes to be processed out of order - the risk is that newer data
+            // could be overwritten by older data.
+            lock (_lock)
```

Comment:
Deadlock? `UpdateCollection` is being called inside a lock on `_lock` in `Initialize`

Reply:
Yeah locks are reentrant YO